runtime/lua/vim/loader.lua

```diff
@@ -27,13 +27,25 @@ M.path = vim.fn.stdpath('cache') .. '/luac'
 ---@nodoc
 M.enabled = false
 
+---@nodoc
+M.charmap = {
```

Comment:
Why is this exposed in the public API? Can it be local?

Reply:
Yeah, I only made it public (and added the reverse lookup) because I thought that users could potentially be able to make use of it. E.g. it provides info on what chars get converted, and could be used via `cachepath:gsub("%%..", vim.loader.charmap)` to derive the original path. Although this could be accomplished without a `vim.loader.charmap` too of course.

Part of what motivated me to put those things in was thinking about how to maintain my own cache on a regular basis. An idea I came up with was to make a script that goes through the cache dir and checks each filename in order to see if its original still exists (and then removing the cached/compiled file if it doesn't).

Lmk what you think. If this is all too far-fetched I can make it local.

Reply:
That is interesting, but then maybe we should expose these existing uri_encode/uri_decode functions and use those instead: https://github.com/neovim/neovim/blob/ccf328172bac2b02f9bd19fa58e105958514a28a/runtime/lua/vim/uri.lua#L19

> Use the same percent-encoding that is done for URLs, but modified to encode a different set of "reserved" characters.

Why is it necessary to have different reserved chars? This prevents using `uri.uri_encode`.

Reply:
> Why is it necessary to have different reserved chars? This prevents using `uri.uri_encode`.

Oh I didn't see that function. I suppose that should probably be used along with the `rfc2396` pattern. The only real downside I can think of would be somewhat longer filenames due to encoding more chars; idk if that would cause or lead to any issues in-practice...probably not?

Reply:
> The only real downside I can think of would be somewhat longer filenames due to encoding more chars

If we are worried about that then we need a different approach altogether, like a compression function :) 

However, this does raise the point that we should raise an error/warning if a platform-specific path limit is exceeded. That could be a followup PR if it is out of scope here.